add `ModuleState`, core lifecycle callbacks for `#[pymodule]`s

bazaah opened the pull request:
This patch set adds the core internal representation for a `PyO3` module's per-module state.

Some of this work is taken from #4162 (credit: @Aequitosh), stripped down.

Within we...

1. Added the core `repr(C)` type that CPython will allocate (space for)
2. Add the requisite hooks in `PyModuleDef` for initialization (`m_slots` Py_mod_exec slot) and freeing (`m_free`)
3. Add a `TypeMap` based store for user state types and expose a limited API to this state via the `PyModuleMethods` trait
4. Test module initialization works as expected

Left for future work:

1. Performance optimizations (we currently unconditionally allocate a `ModuleState` for all modules, even though nothing uses it yet)
2. Very likely, changes to the public state API in `PyModuleMethods` needs work
	- Is `&mut self` valid? I see every other method -- including those that do mutation (`add_*`) -- take `&self` which makes me think this trait is expected handle synchronization internally
	- Notably absent is a method to _delete_ a state type once initialized
	- I'm unsure what happens if two python thread states attempt to access a module's state concurrently... does something (GIL?) synchronize this for us?

bazaah committed:
src/impl_: add pymodule_state for core ModuleState

This commit adds the core ModuleState & StateCapsule structs for
interop with python's ModuleDef m_size field.

We choose to keep the in-interpreter size small, only being a NonNull
ptr to the actual (allocated by Rust) state. This is suboptimal from
a pointer chasing perspective, because we'll eventually have
interpreter->StateCapsule->TypeMap<T> dereference chains, but it makes
it easier to implement and reason about.

bazaah committed:
src/impl_: prepare ModuleDef for per-module state init

This commit sets ModuleDef.m_size to size_of::<ModuleState>, and adds
the relevant callback to ModuleDef.m_free.

We also add C function defs for Py_mod_exec and ModuleDef.m_free
callbacks. These will be used to initialize and free ModuleState in
later patches.

We do not yet set an initializer, as these are passed in m_slots, which
are set from the `#[pymodule]` impl in pyo3-macros-backend.

bazaah committed:
pyo3-macros-backend: add Py_mod_exec slot for initializing ModuleState

bazaah committed:
src/impl_: add test module_state_init

Asserting we initialize ModuleState correctly, that it is accessible
from both module_exec calls -- that is, during module initialization --
and from after the module is loaded.

bazaah committed:
newsfragments: add PR description

davidhewitt reviewed (commented):
Thanks, this seems like a reasonable step forward. The `StateCapsule` type is obviously completely empty, I'd prefer we had some tangible substance before merging.

I don't yet see the full picture of how they state type gets passed around various parts of the PyO3 user code, it would be helpful to understand the current direction and proposed APIs for users to write / read module state.

bazaah committed:
internal/typemap: add TypeMap implementation

This is ported from some of my internal code, with adjustments for
edition = 2024.

It was originally forked from rust-typemap but has diverged
significantly, and been updated for newer rust versions.

I freely relicense it under PyO3's licensing terms.

bazaah committed:
src/types: add module state methods to PyModuleMethods

These commit adds three methods to the external interface for PyModules,
exposing a limited ability to interact with the per-module state area.

- state_ref
- state_mut
- state_or_init

We take a hands-off approach to what the concrete "shape" of this state
is, beyond requiring it is compatible with the marker ModuleStateType
trait.

Modules may store as many types as needed, and all the normal rust
visibility conventions apply, allowing them to guard type invariants, or
prevent external access entirely.

bazaah committed:
src/impl_: switch to real StateMap implementation

Within we switch over to the real typemap backed StateMap
implementation.

We also add a marker trait, ModuleStateType which guarantees the
properties we expect from stored StateMap types.

Namely: Clone + Send (not Sync), though I'm not sure these are the
right guarantees for python's memory model.

Send is a requirement, as python can move PyModules between threads at
will, and Clone makes dealing with ModuleState much easier longer term
(speaking from experience with typemaps).

I'm not sure if Sync should be a requirement too, as it's not clear to
me if PyModules can be concurrently accessed between attached python
thread states. I'm worried the answer might change between free-threaded
and "normal" python builds.

bazaah committed:
src/impl_: add getters for ModuleState->StateMap, from_bound constructors

bazaah committed:
src/impl_: update test to use PyModuleMethods, assert on dummy UserState

bazaah committed:
src/impl_: inline size_of<ModuleState>

bazaah committed:
src/impl_: make ModuleState::drop_impl unsafe, document usage

bazaah committed:
src/impl_: use trampoline in pyo3_module_state_init

So we can return a real error if our state ptr is null.

The upstream cpython builtins just assert non-null, so this is probably
fine

References: https://github.com/python/cpython/blob/v3.13.0/Modules/_sqlite/module.h#L80-L81
References: https://github.com/python/cpython/blob/v3.13.0/Modules/_io/_iomodule.h#L172-L173
References: https://github.com/python/cpython/blob/v3.13.0/Modules/_blake2/blake2module.c#L33-L34

davidhewitt reviewed (commented):
We can definitely have an `experimental-module-state` feature which exposes the public API as we iterate. (We have a couple such features, the goal is to stabilize them when we know they feel good.)

I think I want the API to look something like this:

```rust
#[pymodule]
mod foo {
    #[pymodule_state]
    struct FooState {
        a: Py<PyAny>
    }

    #[pymodule_init]
    fn init(m: &Bound<'_, PyModule>) -> PyResult<FooState> {
        Ok(FooState {
            a: m.py().None()
        })
    }

    #[pyfunction]
    fn foo_func(state: &FooState, py: Python<'_>) -> Py<PyAny> {
        state.a.clone_ref(py)
    }

    #[pyclass]
    struct FooClass { }

    #[pymethods]
    impl FooClass {
        #[new]
        fn new(state: &FooState) -> Self {
            Self { }
        }
    }
}
```

davidhewitt commented:
For module traverse / clear - probably the discussion in #5663 is relevant (could be derived on the module state type).

bazaah commented:
Sorry no @davidhewitt, I'm very busy in the weeks now so I only really get a chance to work on this on the weekends, and I haven't pushed anything relevant yet.
